code/natives/libs/Threader.cxx

```diff
@@ -13,19 +13,59 @@ POSIX::Threader::~Threader() {
     delete this->args;
 }
 
-int detachDispatcher(JavaVM& vm) {
+void attachDispatcher(POSIX::Threader::DispatcherArgs* args) {
+    /* setup jvm thread attach args info */
+    int isAttached = -1;
+
+    JNIEnv* javaEnv;
+    args->javaVM->GetEnv((void**)&(javaEnv), args->jvmArgs->version);
+
+     /* attach the current thread to the Java VM based on the platform */
+    while (isAttached != 0) {
+        #if (Platform == Linux_x64)
+            isAttached = args->javaVM->AttachCurrentThread((void**)&(javaEnv), args->jvmArgs);
+        #elif (Platform == Android_x86_x64)
+            isAttached = args->javaVM->AttachCurrentThread(&(javaEnv), args->jvmArgs);
+        #endif
+    /* wait until the thread is attached to the jvm then start invoking the java funtions */
+    }
+}
+
+void detachDispatcher(JavaVM& vm) {
     int isFinished = -1;
-    
+    int isDestroyed = -1;
+
     while(isFinished != 0) {
         isFinished = vm.DetachCurrentThread();
     }
+    while (isDestroyed != 0) {
+        isDestroyed = vm.DestroyJavaVM();
+    }
+}
 
-    vm.DestroyJavaVM();
+void initMutexOnce(POSIX::Threader::DispatcherArgs* args) {
+    // 0 -> maps a success, non-zero maps a failure (bigger than zero in NPTL API)
+    while (isInitialized != 0) {
+        isInitialized = mutexInit(args->mutex, args->mutexAttr);
+        setMutexAttrType(args->mutexAttr, PTHREAD_MUTEX_DEFAULT);
+    }
+}
+
+void lock(Mutex* mutex) {
+    int isLocked = -1;
+    while (isLocked != 0) {
+        isLocked = lockMutex(mutex);                
+    }
+}
 
-    return isFinished;
+void unlock(Mutex* mutex) {
+    int isUnlocked = -1;
+    while (isUnlocked != 0) {
+        isUnlocked = unlockMutex(mutex);                
+    }
 }
 
-int methodDispatcher(void* arguments) {
+void* methodDispatcher(void* arguments) {
```

Comment:
(1) This is the fix, to include all the actions in single method, which designates a single thread that dispatches java methods using jni reflections.